src/v/iceberg/manifest_entry_values.cc

```diff
@@ -0,0 +1,249 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/manifest_entry_values.h"
+
+#include "bytes/iobuf_parser.h"
+#include "strings/string_switch.h"
+
+namespace iceberg {
+
+namespace {
+
+ss::sstring from_iobuf(iobuf b) {
+    auto size_bytes = b.size_bytes();
+    iobuf_parser parser{std::move(b)};
+    return parser.read_string(size_bytes);
+}
+
+template<typename PrimitiveV>
+auto get_required_primitive(value v) {
+    if (!holds_alternative<primitive_value>(v)) {
+        throw std::invalid_argument("Value is not a primitive");
+    }
+    auto& as_primitive = std::get<primitive_value>(v);
+    if (!holds_alternative<PrimitiveV>(as_primitive)) {
+        throw std::invalid_argument("Value is not the expected type");
+    }
+    // NOTE: values that contain iobufs must be moved.
+    auto& as_t = std::get<PrimitiveV>(as_primitive);
+    return std::move(as_t.val);
+}
+template<typename PrimitiveV>
+auto get_required_primitive(std::optional<value> v) {
+    if (!v.has_value()) {
+        throw std::invalid_argument("Expected primitive value is null");
+    }
+    return get_required_primitive<PrimitiveV>(std::move(*v));
+}
+
+template<typename T, typename PrimitiveV>
+std::optional<T> get_optional_primitive(std::optional<value> v) {
+    if (!v.has_value()) {
+        return std::nullopt;
+    }
+    return T{get_required_primitive<PrimitiveV>(std::move(*v))};
+}
+
+std::unique_ptr<struct_value> get_required_struct(std::optional<value> v) {
+    if (!v.has_value()) {
+        throw std::invalid_argument("Expected struct value is null");
+    }
+    if (!std::holds_alternative<std::unique_ptr<struct_value>>(*v)) {
+        throw std::invalid_argument("Value is not a struct");
+    }
+    auto ret = std::get<std::unique_ptr<struct_value>>(std::move(*v));
+    if (!ret) {
+        throw std::invalid_argument("Struct value is nullptr");
```

Comment:
Do we expect this to happen normally or is this defense against bugs? I don't have a problem with adding in extra checks, but it might be worth a comment if that's the case.

Reply:
The values we have here _should_ abide by the schema defined by the manifest entry type. These are all defensive checks.